File: sdk_v2/js/src/openai/audioClient.ts
// `AudioClient`, layered on `AudioSession` via the OpenAI-JSON pass-through pattern. Mirrors C# `OpenAIAudioClient`.

import { Item, type Item as ItemT, type TextItem } from "../items.js";
import type { Model } from "../model.js";
import { Request } from "../request.js";
import type { Response } from "../response.js";
import { AudioSession } from "../session.js";
import { LiveAudioTranscriptionSession } from "./liveAudioSession.js";

// biome-ignore lint/suspicious/noExplicitAny: OpenAI request/response objects are user-shaped JSON.
type Json = any;

export class AudioClientSettings {
  language?: string;
  temperature?: number;

  /** @internal */
  _serialize(): Record<string, Json> {
    const result: Record<string, Json> = {};
    if (this.language !== undefined) result.language = this.language;
    if (this.temperature !== undefined) result.temperature = this.temperature;
    return result;
  }
}

export class AudioClient {
  readonly #model: Model;
  #session: AudioSession | undefined;

  public settings = new AudioClientSettings();

  /** @internal — construct via `model.createAudioClient()`. */
  constructor(model: Model) {
    this.#model = model;
  }

  get modelId(): string {
    return this.#model.id;
  }

  #ensureSession(): AudioSession {
    if (this.#session === undefined) {
      this.#session = new AudioSession(this.#model);
    }
    return this.#session;
  }

  /**
   * Transcribe an audio file. Path is forwarded to native and may be any
   * format the underlying model accepts.
   */
  async transcribe(filePath: string): Promise<Json> {
    if (typeof filePath !== "string" || filePath.trim() === "") {
      throw new Error("filePath must be a non-empty string.");
    }
    const requestJson: Record<string, Json> = {
      model: this.modelId,
      filename: filePath,
      ...this.settings._serialize(),
    };

    const session = this.#ensureSession();
    const request = new Request();
    request.addItem(Item.text(JSON.stringify(requestJson), "openai-json"));

    let response: Response;
    try {
      response = await session.processRequest(request);
    } catch (err) {
      throw new Error(
        `Audio transcription failed for model '${this.modelId}': ${err instanceof Error ? err.message : String(err)}`,
        { cause: err },
      );
    }

    const text = findOpenAiJsonText(response.output);
    if (text === undefined) {
      throw new Error(
        `Audio transcription for model '${this.modelId}' returned no openai-json text item.`,
      );
    }
    return JSON.parse(text);
  }

  /** Streaming transcribe. Yields parsed JSON chunks as the model emits them. */
  transcribeStreaming(filePath: string): AsyncIterable<Json> {
    if (typeof filePath !== "string" || filePath.trim() === "") {
      throw new Error("filePath must be a non-empty string.");
    }
    const requestJson: Record<string, Json> = {
      model: this.modelId,
      filename: filePath,
      stream: true,
      ...this.settings._serialize(),
    };

    const session = this.#ensureSession();
    const modelId = this.modelId;
    return {
      async *[Symbol.asyncIterator](): AsyncIterator<Json> {
        const request = new Request();
        request.addItem(Item.text(JSON.stringify(requestJson), "openai-json"));

        try {
          for await (const item of session.processStreamingRequest(request)) {
            if (item.type !== "text") continue;
            const t = item as TextItem;
            if (t.textType !== "openai-json" || t.text === "") continue;
            yield JSON.parse(t.text);
          }
        } catch (err) {
          if (err instanceof Error && err.name === "AbortError") throw err;
          throw new Error(
            `Streaming audio transcription failed for model '${modelId}': ${err instanceof Error ? err.message : String(err)}`,
            { cause: err },
          );
        }
      },
    };
  }

  /**
   * Create a real-time audio transcription session for streaming PCM input.
   * Call `start()`, push PCM chunks via `append(chunk)`, consume parsed
   * transcription responses via `getStream()`, then `stop()` and `dispose()`.
   */
  createLiveTranscriptionSession(): LiveAudioTranscriptionSession {
    return new LiveAudioTranscriptionSession(this.#model);
  }
}

function findOpenAiJsonText(output: ReadonlyArray<ItemT>): string | undefined {
  for (const item of output) {
    if (item.type === "text" && (item as TextItem).textType === "openai-json") {
      return (item as TextItem).text;
    }
  }
  return undefined;
}

File: sdk_v2/rust/src/openai/audio_client.rs
//! OpenAI-compatible audio transcription client.

use std::path::Path;
use std::pin::Pin;
use std::sync::Arc;
use std::task::{Context, Poll};

use serde_json::{json, Value};

use crate::detail::core_interop::CoreInterop;
use crate::error::{FoundryLocalError, Result};

/// OpenAI-compatible audio transcription response.
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct AudioTranscriptionResponse {
    /// The transcribed text.
    pub text: String,
    /// The language of the input audio (if detected).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    /// Duration of the input audio in seconds (if available).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration: Option<f64>,
    /// Segments of the transcription (if available).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub segments: Option<Vec<serde_json::Value>>,
    /// Words with timestamps (if available).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub words: Option<Vec<serde_json::Value>>,
}

/// Tuning knobs for audio transcription requests.
///
/// Use the chainable setter methods to configure, e.g.:
///
/// ```ignore
/// let mut client = model.create_audio_client();
/// client.language("en").temperature(0.2);
/// ```
#[derive(Debug, Clone, Default)]
pub struct AudioClientSettings {
    language: Option<String>,
    temperature: Option<f64>,
}

impl AudioClientSettings {
    /// Serialise settings into the JSON fragment expected by the native core.
    fn serialize(&self, model_id: &str, file_name: &str) -> Value {
        let mut map = serde_json::Map::new();

        map.insert("Model".into(), json!(model_id));
        map.insert("FileName".into(), json!(file_name));

        if let Some(ref lang) = self.language {
            map.insert("Language".into(), json!(lang));
        }
        if let Some(temp) = self.temperature {
            map.insert("Temperature".into(), json!(temp));
        }

        Value::Object(map)
    }
}

/// A stream of [`AudioTranscriptionResponse`] chunks.
///
/// Returned by [`AudioClient::transcribe_streaming`].
pub struct AudioTranscriptionStream {
    rx: tokio::sync::mpsc::UnboundedReceiver<Result<String>>,
}

impl futures_core::Stream for AudioTranscriptionStream {
    type Item = Result<AudioTranscriptionResponse>;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        match self.rx.poll_recv(cx) {
            Poll::Ready(Some(Ok(chunk))) => {
                if chunk.is_empty() {
                    cx.waker().wake_by_ref();
                    Poll::Pending
                } else {
                    let parsed = serde_json::from_str::<AudioTranscriptionResponse>(&chunk)
                        .map_err(FoundryLocalError::from);
                    Poll::Ready(Some(parsed))
                }
            }
            Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(e))),
            Poll::Ready(None) => Poll::Ready(None),
            Poll::Pending => Poll::Pending,
        }
    }
}

/// Client for OpenAI-compatible audio transcription backed by a local model.
pub struct AudioClient {
    model_id: String,
    core: Arc<CoreInterop>,
    settings: AudioClientSettings,
}

impl AudioClient {
    pub(crate) fn new(model_id: String, core: Arc<CoreInterop>) -> Self {
        Self {
            model_id,
            core,
            settings: AudioClientSettings::default(),
        }
    }

    /// Set the language hint for transcription.
    pub fn language(&mut self, lang: impl Into<String>) -> &mut Self {
        self.settings.language = Some(lang.into());
        self
    }

    /// Set the sampling temperature.
    pub fn temperature(&mut self, v: f64) -> &mut Self {
        self.settings.temperature = Some(v);
        self
    }

    /// Transcribe an audio file.
    pub async fn transcribe(
        &self,
        audio_file_path: impl AsRef<Path>,
    ) -> Result<AudioTranscriptionResponse> {
        let path_str =
            audio_file_path
                .as_ref()
                .to_str()
                .ok_or_else(|| FoundryLocalError::Validation {
                    reason: "audio file path is not valid UTF-8".into(),
                })?;
        Self::validate_path(path_str)?;

        let request = self.settings.serialize(&self.model_id, path_str);
        let params = json!({
            "Params": {
                "OpenAICreateRequest": serde_json::to_string(&request)?
            }
        });

        let raw = self
            .core
            .execute_command_async("audio_transcribe".into(), Some(params))
            .await?;
        let parsed: AudioTranscriptionResponse = serde_json::from_str(&raw)?;
        Ok(parsed)
    }

    /// Transcribe an audio file with streaming results, returning an
    /// [`AudioTranscriptionStream`].
    pub async fn transcribe_streaming(
        &self,
        audio_file_path: impl AsRef<Path>,
    ) -> Result<AudioTranscriptionStream> {
        let path_str =
            audio_file_path
                .as_ref()
                .to_str()
                .ok_or_else(|| FoundryLocalError::Validation {
                    reason: "audio file path is not valid UTF-8".into(),
                })?;
        Self::validate_path(path_str)?;

        let request = self.settings.serialize(&self.model_id, path_str);

        let params = json!({
            "Params": {
                "OpenAICreateRequest": serde_json::to_string(&request)?
            }
        });

        let rx = self
            .core
            .execute_command_streaming_channel("audio_transcribe".into(), Some(params))
            .await?;

        Ok(AudioTranscriptionStream { rx })
    }

    fn validate_path(path: &str) -> Result<()> {
        if path.trim().is_empty() {
            return Err(FoundryLocalError::Validation {
                reason: "audio_file_path must be a non-empty string".into(),
            });
        }
        Ok(())
    }
}
